AnnotatedImageAdapter for Image file plus additional JSON data

ecrows opened the pull request:
## Description
Further to [Issue #934](https://github.com/bentoml/BentoML/issues/934)

Added an AnnotatedImageInput adapter that uses multipart/form-data to read in an image file and JSON file at the same time.  This is to handle the common use-case for models that require additional information about an image (such as object detection bounding boxes, segmentation masks, class labels, face keypoints, etc.) 

Unit tests and documentation are included!

## Motivation and Context
See the conversation in [the issue here](https://github.com/bentoml/BentoML/issues/934)

A common use-case for models that take image input is to also include additional information about the image (such as object detection bounding boxes, segmentation masks, class labels, face keypoints, etc.) ImageInput does not allow you to include this kind of information.

## How Has This Been Tested?
Wrote and ran unit tests under all environments using tox.  Successfully using the AnnotatedImageAdapter for my own project for work.

## Types of changes
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [x] New feature and improvements (non-breaking change which adds/improves functionality)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Code Refactoring (internal change which is not user facing)
- [x] Documentation
- [x] Test, CI, or build

## Component(s) if applicable
- [x] BentoService (service definition, dependency management, API input/output adapters)
- [ ] Model Artifact (model serialization, multi-framework support)
- [ ] Model Server (micro-batching, dockerisation, logging, OpenAPI, instruments)
- [ ] YataiService gRPC server (model registry, cloud deployment automation)
- [ ] YataiService web server (nodejs HTTP server and web UI)
- [ ] Internal (BentoML's own configuration, logging, utility, exception handling)
- [ ] BentoML CLI

## Checklist:
- [x] My code follows the bentoml code style, both `./dev/format.sh` and
  `./dev/lint.sh` script have passed
  ([instructions](https://github.com/bentoml/BentoML/blob/master/DEVELOPMENT.md#style-check-and-auto-formatting-your-code)).
- [x] My change reduces project test coverage and requires unit tests to be added
- [x] I have added unit tests covering my code change
- [x] My change requires a change to the documentation
- [x] I have updated the documentation accordingly

Thanks for reviewing my changes!  As far as areas of concern may go, I'm not sure if my allowance of `is_batch_input` is correct, or if I should instead be raising an error if this is set to True.  I'm not that familiar with AWS lambda (I modeled the corresponding unit tests after MultiImageInput), so that may also be of interest.  Finally, I've been running into some errors stemming from an undefined BentoService repository path after rebasing and running tox (which I don't think these are related to my changes, but I want to be sure).

Thanks again, let me know what needs to be fixed up!
-Evan

ecrows commented:
I see from codecov that my code coverage is only around 75% at the moment.  I can spend some time early next week to bring that up above 90%.

yubozhao commented:
@ecrows This looks great. Good job adding it to the API page on the documentation site.

pep8speaks commented:
Hello @ecrows, Thanks for updating this PR.



* In the file [`bentoml/adapters/annotated_image_input.py`](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/bentoml/adapters/annotated_image_input.py):

> [Line 256:5](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/bentoml/adapters/annotated_image_input.py#L256): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (2)
> [Line 288:17](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/bentoml/adapters/annotated_image_input.py#L288): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 288:48](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/bentoml/adapters/annotated_image_input.py#L288): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator





* In the file [`tests/adapters/test_annotated_image_handler.py`](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/tests/adapters/test_annotated_image_handler.py):

> [Line 66:13](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/tests/adapters/test_annotated_image_handler.py#L66): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 104:1](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/tests/adapters/test_annotated_image_handler.py#L104): [E303](https://duckduckgo.com/?q=pep8%20E303) too many blank lines (3)
> [Line 149:64](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/tests/adapters/test_annotated_image_handler.py#L149): [E226](https://duckduckgo.com/?q=pep8%20E226) missing whitespace around arithmetic operator
> [Line 215:13](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/tests/adapters/test_annotated_image_handler.py#L215): [E128](https://duckduckgo.com/?q=pep8%20E128) continuation line under-indented for visual indent
> [Line 315:62](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/tests/adapters/test_annotated_image_handler.py#L315): [W504](https://duckduckgo.com/?q=pep8%20W504) line break after binary operator
> [Line 334:56](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/tests/adapters/test_annotated_image_handler.py#L334): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 334:58](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/tests/adapters/test_annotated_image_handler.py#L334): [E251](https://duckduckgo.com/?q=pep8%20E251) unexpected spaces around keyword / parameter equals
> [Line 358:13](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/tests/adapters/test_annotated_image_handler.py#L358): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 362:25](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/tests/adapters/test_annotated_image_handler.py#L362): [E711](https://duckduckgo.com/?q=pep8%20E711) comparison to None should be 'if cond is None:'
> [Line 363:38](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/tests/adapters/test_annotated_image_handler.py#L363): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 364:13](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/tests/adapters/test_annotated_image_handler.py#L364): [E127](https://duckduckgo.com/?q=pep8%20E127) continuation line over-indented for visual indent
> [Line 365:25](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/tests/adapters/test_annotated_image_handler.py#L365): [E711](https://duckduckgo.com/?q=pep8%20E711) comparison to None should be 'if cond is None:'
> [Line 368:13](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/tests/adapters/test_annotated_image_handler.py#L368): [E126](https://duckduckgo.com/?q=pep8%20E126) continuation line over-indented for hanging indent
> [Line 370:29](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/tests/adapters/test_annotated_image_handler.py#L370): [E711](https://duckduckgo.com/?q=pep8%20E711) comparison to None should be 'if cond is None:'
> [Line 471:9](https://github.com/bentoml/BentoML/blob/acdba1ce428e9e785976030bd6efd99268664400/tests/adapters/test_annotated_image_handler.py#L471): [F841](https://duckduckgo.com/?q=pep8%20F841) local variable 'test_anno_image_input' is assigned to but never used

ecrows commented:
I've added more unit tests, and coverage is up to 100%.  Not a guarantee that nothing will go wrong, but not a bad start. :)

The adapter is working fine for me in my own project, but let me know if there's anything else that you want me to look at before merging.

parano commented:
Thanks for the PR @ecrows and for taking the effort to bring up the test coverage! I will help review.

cc @bojiang @akainth015 to take a quick look and see if there's anything that may affect #982

ecrows commented on bentoml/adapters/annotated_image_input.py:
```diff
@@ -0,0 +1,372 @@
+
+    Example:
+
+        >>> python
```
Thanks, fixed!

ecrows commented on bentoml/adapters/annotated_image_input.py:
```diff
@@ -0,0 +1,372 @@
+    return False
+
+
+def has_image_extension(file_name, accept_format_list):
```
I think that's a good idea, I'll get on that tomorrow.

ecrows replied:
Alright!  I've added type hints for these in my latest commit.  I've also made `accept_format_list` required for both the `has_image_extension` and `verify_image_format_or_raise` functions.  If unset during `__init__`, it is set to the default value.  If the downstream code manages to change it to "None" sometime after object construction, the code will crash when trying to check image types (rather than silently defaulting back the default values).

This seemed reasonable to me, as messing with the accept_image_format_list after construction is unusual, and the caller should explicitly specify if they want to use the defaults.  It isn't clear (in my mind) whether `None` should imply "use the default" or should imply "an empty list where no images are accepted".

If you'd rather I changed it so that it gracefully interprets "None" as "use the defaults", that's no problem at all.

ecrows commented:
Thanks for the feedback so far (feel free to keep going!).  I'll make some more changes tomorrow, and rebase onto the latest version of master.

ecrows commented:
I've rebased onto the latest version and added optional input name arguments for the image and annotation files, which are reflected in the SwaggerUI schema and keyword arguments.  Feel free to take a look whenever you've got a minute.

parano commented on bentoml/adapters/annotated_image_input.py:
```diff
@@ -0,0 +1,384 @@
+        super(AnnotatedImageInput, self).__init__(
+            is_batch_input=is_batch_input, **base_kwargs
+        )
+        if 'input_names' in base_kwargs:
```
It is ok to remove l156-l160,  `input_names` check was in the ImageInput because of the older ImageHandler API supports multiple image input, which was moved to the MultiImageInput adapter, and it was meant for a warning for users using the older API.

ecrows replied:
Removed in the latest version!

parano commented on bentoml/adapters/annotated_image_input.py:
```diff
@@ -0,0 +1,384 @@
+        >>> @artifacts([TensorflowArtifact('classifer')])
+        >>> class PetClassification(BentoService):
+        >>>    @api(input=AnnotatedImageInput())
+        >>>    def predict(self, image, annotations):
```
nit:  rename parameter from image to images,  to indicate that it is a list of images?

might be best to add type annotation here in this example code as well, e.g.:

```def predict(self, images: List[Numpy.Array], annotations: List[JsonSerializable]):```

ecrows replied:
Currently the endpoint takes a single image and a single file for annotations -- I can add type hints that reflect this!

ecrows replied:
Added type hints to reflect what files are expected.

parano commented on bentoml/adapters/annotated_image_input.py:
```diff
@@ -0,0 +1,384 @@
+        >>>    def predict(self, image, annotations):
+        >>>        cropped_pets = some_pet_finder(image, annotations)
+        >>>        results = self.artifacts.classifer.predict(cropped_pets)
+        >>>        return [CLASS_NAMES[r] for r in results]
```
could you also add a client API request example for this example service here?  A curl command or python code that makes a sample HTTP request to this service would work.  here's an example: https://github.com/bentoml/BentoML/blob/v0.8.5/bentoml/adapters/json_input.py

ecrows replied:
Absolutely, will do!

ecrows replied:
Added a HTTP and HTML form section to the docstring (similar to MultiImageInput)

parano commented on bentoml/adapters/image_input.py:
````diff
@@ -81,20 +81,18 @@ class ImageInput(BaseInputAdapter):
-                results = self.artifacts.classifer.predict(image_ndarrays)
-                return [CLASS_NAEMS[r] for r in results]
-        ```
+        >>> from bentoml import BentoService, api, artifacts
````
👍

bojiang commented on bentoml/adapters/annotated_image_input.py:
```diff
@@ -0,0 +1,384 @@
+    try:
+        parsed = json.load(json_file)
+    except (json.JSONDecodeError, UnicodeDecodeError):
+        raise BadInput("BentoML#AnnotatedImageInput received invalid JSON file")
```
👍

bojiang commented:
LGTM

------

After looking through this PR, I feel even more that the current adapter logic is kinda cumbersome. The good news is that we(@akainth015 and me) are working on optimizing the logic here to reduce the burden of thinking when developing adapters.

bojiang commented on bentoml/adapters/annotated_image_input.py:
```diff
@@ -0,0 +1,384 @@
+        result = func(**input_data)
+        return self.output_adapter.to_response(result, request)
+
+    def handle_cli(self, args, func):
```
This API `handle_cli` may change in the next version, adding support for batch input and other common logic for each adapter.

ecrows replied:
Got it, thanks for the heads up.

ecrows commented:
Thanks for all the great feedback folks.  I've made revisions accordingly, but if there's any details you still want addressed or anything that isn't quite resolved, just let me know.

parano commented:
Thanks again @ecrows, everything looks great! Merging now!
